src/soc/sw/baremetal-libs/src/libbaremetal/util.c

```diff
@@ -38,22 +41,34 @@ int optimsoc_get_ctrank(void) {
 
 
 int optimsoc_get_tilerank(unsigned int tile) {
-    uint16_t *ctlist = (uint16_t*) OPTIMSOC_NA_CT_LIST;
     for (int i = 0; i < optimsoc_get_numct(); i++) {
-        if (REG16(&ctlist[i]) == tile) {
+        if (ctlist[i] == tile) {
             return i;
         }
     }
     return -1;
 }
 
 int optimsoc_get_ranktile(unsigned int rank) {
-    uint16_t *ctlist = (uint16_t*) OPTIMSOC_NA_CT_LIST;
     return ctlist[rank];
 }
 
 void optimsoc_init(optimsoc_conf *config) {
-
+  uint32_t numct = REG32(OPTIMSOC_NA_CT_NUM);
+  uint32_t numtiles = REG32(OPTIMSOC_NA_NUMTILES);
+  uint32_t *nalist = (uint32_t*) OPTIMSOC_NA_CT_LIST;
+
+  ctlist = malloc(numct * sizeof(uint16_t));
+  uint32_t idx = 0;
+
+  for (int g = 0; g < ((numtiles+31)>>5); ++g) {
```

Comment:
Could you add a comment here what this code does? (Or even better: make it a separate small function with a descriptive name.)

Reply:
thanks